File: algorithm/回溯算法/preorder_traversal_i_compact.java
package algorithm.回溯算法;

import dataStructure.树.TreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static dataStructure.Util.PrintUtil.printTree;

/**
 *
 * 给定一棵二叉树，搜索并记录所有值为7的节点，请返回节点列表
 *
 */
public class preorder_traversal_i_compact {

    static List<TreeNode> res = new ArrayList<>();

    // 前序遍历
    public static void preOrder(TreeNode root) {
        if (root == null) return;
        if (root.val == 7) res.add(root);
        preOrder(root.left);
        preOrder(root.right);
    }

    /**
     *
     *
     *         /——— 7
     *     /——— 3
     *    |    \——— 6
     * ——— 1
     *    |    /——— 5
     *     \——— 7
     *         \——— 4
     *
     *
     *
     *
     */
    public static void main(String[] args) {
        TreeNode root = new TreeNode(new ArrayList<>(Arrays.asList(1,7,3,4,5,6,7)));
        preOrder(root);
        for (TreeNode node : res) {
            printTree(node);
        }
    }

}

File: 测试代码/preorder_traversal_i_compact.ts
import { TreeNode } from 'dataStructure/树/TreeNode';

// 给定一棵二叉树，搜索并记录所有值为7的节点，返回节点列表

// 前序遍历
function preOrder(root: TreeNode | null, res: TreeNode[]): void {
  if(root === null) {
    return;
  }
  if(root.val === 7) {
    res.push(root);
  }
  preOrder(root.left, res);
  preOrder(root.right, res);
}


// 在二叉树中搜索所有值为7的节点，请返回根节点到这些节点的路径
function preOrder2(
  root: TreeNode | null,
  path: TreeNode[],
  res: TreeNode[][]
): void {
  if(root === null) {
    return;
  }
  // 尝试
  path.push(root);
  if(root.val === 7) {
    // 记录解
    res.push([...path]);
  }
  preOrder2(root.left, path, res);
  preOrder2(root.right, path, res);
  // 回退
  path.pop();
}


// 在二叉树中搜索所有值为7的节点，请返回根节点到这个节点的路径，并要求路径中不包含3的节点
function preOrder3(
  root: TreeNode | null,
  path: TreeNode[],
  res: TreeNode[][]
): void {
  // 剪枝
  if(root === null || root.val === 3) {
    return;
  }
  // 尝试
  path.push(root);
  if(root.val === 7) {
    // 记录解
    res.push([...path]);
  }
  preOrder3(root.left, path, res);
  preOrder3(root.right, path, res);
  // 回退
  path.pop();
}
